### PR TITLE
AssignmentNotUsedInspection false positives - incorrect variable scoping and others

### LINKED ISSUE #4754: Collection of false positives for inspection: "Assignment is not used"
I recently upgraded from 2.3.1.4441 to 2.4.0.4488 and I'm reasonably confident that these false positives haven't been showing up before the upgrade.

For reference, the inspection tells me: *An assignment is immediately overridden by another assignment or is never used.*  
I hope upon inspecting the following procedures you will agree with me that the above assessment is objectively false. Also please disregard their mostly nonsensical nature - I tried to condense them as far as possible. And please don't try to actually execute them.  ;)

``` vb
Option Explicit

Public Function IssueInIfBlock(ByVal someInput As Boolean) As Boolean
    
    If someInput Then
        Dim HasProblems As Boolean
        HasProblems = True ' triggers "Assignment is not used" inspection
    End If
    
    IssueInIfBlock = someInput And HasProblems
    
End Function

Public Function IssueInSelectStatement(ByVal someInput As Long) As Boolean
        
    Select Case someInput
        Case True
            Dim HasProblems As Boolean
            HasProblems = True ' triggers "Assignment is not used" inspection
    End Select
    
    IssueInSelectStatement = someInput And HasProblems
    
End Function

Public Function IssueWithStaticVariable() As Boolean

    Static SomeProperty As String
    If SomeProperty = "dummy" Then
        IssueWithStaticVariable = True
    End If
    SomeProperty = "new dynamic value" ' triggers "Assignment is not used" inspection
    
End Function

Public Sub IssueInWithBlock()
    
    Dim SomeObject As Object
    With SomeObject
        Dim HasProblems As Boolean
        HasProblems = SomeObject.SomeProperty ' triggers "Assignment is not used" inspection
    End With
    Debug.Print HasProblems
    
End Sub

Public Sub IssueInDoLoop()

    Dim Idx As Long
    Idx = 1
    Do
        Dim SomeBreakCondition As Boolean
        SomeBreakCondition = True ' triggers "Assignment is not used" inspection
        Idx = Idx + 1
    Loop Until SomeBreakCondition
    
End Sub
```

Edit: As detailed in the comments, moving the `Dim` statements before the blocks where their variables appear in will make the inspection disappear. Only the inspection for `IssueWithStaticVariable` will remain.

### LINKED ISSUE #4913: False Positive 'Assignment is not used' inspection
**Rubberduck version information**
Version 2.4.1.4661
OS: Microsoft Windows NT 10.0.15063.0, x64
Host Product: Microsoft Office 2016 x64
Host Version: 16.0.4822.1000
Host Executable: EXCEL.EXE

**Description**
False positive `Assignment not used` inspection on the following use cases. 
The actual code where it surfaced:
```VBA
  Dim sqlString As String
  If AllClinicFlag Then
    sqlString = selectStatementBase & Table & " WHERE " & dateColumnName & " BETWEEN '" & startDate & "' AND '" & endDate & "'"
  Else
    sqlString = selectStatementBase & Table & whereClause & clinicID
    sqlString = sqlString & " AND " & dateColumnName & " BETWEEN '" & startDate & "' AND '" & endDate & "'"
  End If
  If Len(Extra) > 0 Then
    sqlString = sqlString & " AND " & Extra
  End If
  
  LogManager.Log TraceLevel, "SQLString: " & sqlString
```
An MCVE
```VBA
  Dim foo As String
  foo = "bar"
  foo = foo & " baz"
```

**To Reproduce**
1. See above

**Expected behavior**
no inspection necessary in these cases

**Screenshots**
n/a

**Logfile**
[RubberduckLog.txt](https://github.com/rubberduck-vba/Rubberduck/files/3075000/RubberduckLog.txt)

**Additional context**
I had quite a number of other legitimate, valid `Assignment is not used` inspection results, so it's not totally borked.

Possibly related: #4754 

_Yes, I know, Bobby Tables... It's on the list of refactorings to do. I'm the only one in the code and the only user. Since I have no intention of causing self-harm, it's_ low _on the list to fix_

### PR DESCRIPTION
**Pull request by BZngr**
Closes #4913 
Closes #4754 

1. Removes 'block-based' evaluations.
2. Adds support for `static` variables and variables within static members.
3. Correctly handles prior assignments referenced by a subsequent assignment, e.g.,
```
Function Test() As String
    Dim fizz As String
    fizz = "hello" '<--- this assignment previously flagged as unused'
    fizz = fizz & " world"
    Test = fizz
End Function
```

### CODE REVIEW

**Comment by MDoerner** [comment]
Thinking a bit more about the `Resume` case, could you please add a test with a single overall error handler, whose OnErrorStatement is at the start of the method. Then there should be two assignments, one used and one not used. This should still return a result. I think, the way the code is set upright now, it will not.

File: Rubberduck.CodeAnalysis/Inspections/Concrete/AssignmentNotUsedInspection.cs [hunk 1]
```diff
@@ -58,54 +73,101 @@ protected override IEnumerable<IdentifierReference> ReferencesInModule(Qualified
...
+            var tree = walker.GenerateTree(localVariable.ParentScopeDeclaration.Context, localVariable);
+
+            var allAssignmentsAndReferences = tree.Nodes(new[] { typeof(AssignmentNode), typeof(ReferenceNode) })
+                                    .Where(node => localVariable.References.Contains(node.Reference));
+
+            var unusedAssignmentNodes = allAssignmentsAndReferences.Any(n => n is ReferenceNode)
+                        ? FindUnusedAssignmentNodes(tree, localVariable, allAssignmentsAndReferences)
+                        : allAssignmentsAndReferences.OfType<AssignmentNode>();
+
+            return unusedAssignmentNodes.Except(FindDescendantsOfNeverFlagNodeTypes(unusedAssignmentNodes))
```
[the remark this answers is not in the document]
Comment (BZngr) [reply Rubberduck.CodeAnalysis/Inspections/Concrete/AssignmentNotUsedInspection.cs]: Yes - clearer and results in a few less lines of code.

File: Rubberduck.CodeAnalysis/Inspections/Concrete/AssignmentNotUsedInspection.cs [hunk 2]
```diff
@@ -58,54 +73,101 @@ protected override IEnumerable<IdentifierReference> ReferencesInModule(Qualified
...
                 }
             }
 
-            return nodes;
+            return allAssignmentsAndReferences.OfType<AssignmentNode>()
+                                                .Except(usedAssignments);
         }
 
-        protected override bool IsResultReference(IdentifierReference reference, DeclarationFinder finder)
+        private static IEnumerable<AssignmentNode> FindDescendantsOfNeverFlagNodeTypes(IEnumerable<AssignmentNode> flaggedAssignments)
```
[the remark this answers is not in the document]
Comment (BZngr) [reply Rubberduck.CodeAnalysis/Inspections/Concrete/AssignmentNotUsedInspection.cs]: This method is now deleted and replaced by `IsDescendentOfNeverFlagNode`...though it definitely had the issues you pointed out.

File: Rubberduck.CodeAnalysis/Inspections/Concrete/AssignmentNotUsedInspection.cs [hunk 3]
```diff
@@ -131,31 +193,21 @@ private static bool IsPotentiallyUsedViaJump(IdentifierReference resultCandidate
...
-            }
-
-            return false;
-        }
-
-        private static bool TryGetRelevantJumpContext<T>(IdentifierReference resultCandidate, out T ctxt) where T : ParserRuleContext //, IEnumerable<T> stmtContexts, int targetLine, int? targetColumn = null) where T : ParserRuleContext
+        private static bool TryGetRelevantJumpContext<T>(IdentifierReference resultCandidate, out T ctxt) where T : ParserRuleContext
         {
             ctxt = resultCandidate.ParentScoping.Context.GetDescendents<T>()
                                     .Where(sc => sc.Start.Line > resultCandidate.Context.Start.Line
```
[the remark this answers is not in the document]
Comment (BZngr) [reply Rubberduck.CodeAnalysis/Inspections/Concrete/AssignmentNotUsedInspection.cs]: I have changed it to use `Selection` for comparisons but had to modify the suggested code slightly:
```
            ctxt = resultCandidate.ParentScoping.Context.GetDescendents<T>()
                                    .Where(descendent => descendent.GetSelection() > resultCandidate.Selection)
                                    .OrderBy(descendent => descendent.GetSelection())
                                    .FirstOrDefault();

```

**Comment by BZngr** [comment]
@MDoerner First, thanks for the awesome review.  I have updated all the code sections per your comments.  If I did not directly reply to the code comment - you can assume the answer is 'Done'.

Regarding the initial general comments:
1. Line 231 concern: changed to a `switch` statement as recommended.
2. Line numbers: yes - original implementation missed the mark entirely.  The correction was pretty straight forward though.  Thanks for spotting this.
3. `Resume` is considered in the analysis to reduce false positives from used assignments within the error handler section.
4.  The remaining points are valid as well - and, are left as a conceded analysis gaps until they can be addressed using improved code path analysis.

I added the following test per your last comment above:
```
        [Test]
        public void ResumeStmtTest_FromComments()
        {
            string code =
$@"
Public Function Inverse(value As Double) As Double
On Error GoTo ErrorHandler:
    Dim ratio As Double
    ratio = 0# '<== unused
    ratio = 1# / value '<== used
    Inverse = ratio
    Exit Function
ErrorHandler: 
    ratio = 0#  '<== possibly used
    Resume
End Function
";
            var results = InspectionResultsForStandardModule(code);
            Assert.AreEqual(1, results.Count());
        }

```

The test passed without changes to code.  So, if this is the test you had in mind, I do not intend to add it to the code base.

**Pull request merged**